File: 2019/intcode.jl
module Intcode

export State, make_computer
export initialize, load_program, add_input, execute, get_output
export get_state, read_memory, get_instruction_pointer

@enum State IDLE RUN WAIT HALT

mutable struct Computer
    ram::Dict{Int, Int}
    state::State
    instruction_pointer::Int
    relative_base::Int
    inputs::Array{Int, 1}
    outputs::Array{Int, 1}
end

function make_computer()
    Computer(Dict(), IDLE, 0, 0, Int[], Int[])
end

function initialize(computer::Computer)
    computer.state = RUN
end

function load_program(computer::Computer, program::Dict{Int, Int})
    computer.ram = program
    computer.instruction_pointer = 0
    computer.relative_base = 0
    computer.inputs = Int[]
    computer.outputs = Int[]
end

function add_input(computer::Computer, in_val::Int)
    push!(computer.inputs, in_val)
end

function get_instruction_pointer(computer::Computer)::Int
    computer.instruction_pointer
end

function get_state(computer::Computer)::State
    computer.state
end

function get_output(computer::Computer)::Int
    pop!(computer.outputs)
end

function read_memory(computer::Computer, address::Int)::Int
    computer.ram[address]
end

function execute(computer::Computer)
    instruction = get!(computer.ram, computer.instruction_pointer, 0)
    opcode = instruction % 100
    instruction ÷= 100

    in1_mode = instruction % 10
    in1_pos = Int(0)
    if in1_mode == 0
        in1_pos = get!(computer.ram, computer.instruction_pointer + 1, 0)
    elseif in1_mode == 1
        in1_pos = computer.instruction_pointer + 1
    elseif in1_mode == 2
        in1_pos = get!(computer.ram, computer.instruction_pointer + 1, 0) + computer.relative_base
    else
        println("Invalid input 1 mode!")
    end
    instruction ÷= 10

    in2_mode = instruction % 10
    in2_pos = 0
    if in2_mode == 0
        in2_pos = get!(computer.ram, computer.instruction_pointer + 2, 0)
    elseif in2_mode == 1
        in2_pos = computer.instruction_pointer + 2
    elseif in2_mode == 2
        in2_pos = get!(computer.ram, computer.instruction_pointer + 2, 0) + computer.relative_base
    else
        println("Invalid input 2 mode!")
    end
    instruction ÷= 10

    out_mode = instruction % 10
    out_pos = 0
    if out_mode == 0
        out_pos = get!(computer.ram, computer.instruction_pointer + 3, 0)
    elseif out_mode == 1
        out_pos = computer.instruction_pointer + 3
    elseif out_mode == 2
        out_pos = get!(computer.ram, computer.instruction_pointer + 3, 0) + computer.relative_base
    else
        println("Invalid output mode!")
    end

    if opcode == 99
        halt(computer)
    elseif opcode == 1
        add(computer, in1_pos, in2_pos, out_pos)
    elseif opcode == 2
        multiply(computer, in1_pos, in2_pos, out_pos)
    elseif opcode == 3
        inp(computer, in1_pos)
    elseif opcode == 4
        output(computer, in1_pos)
    elseif opcode == 5
        jump_if_true(computer, in1_pos, in2_pos)
    elseif opcode == 6
        jump_if_false(computer, in1_pos, in2_pos)
    elseif opcode == 7
        less_than(computer, in1_pos, in2_pos, out_pos)
    elseif opcode == 8
        equals(computer, in1_pos, in2_pos, out_pos)
    elseif opcode == 9
        offset_relative_base(computer, in1_pos)
    else
        println("Invalid opcode!")
    end
end

function halt(computer::Computer)
    computer.instruction_pointer += 1
    computer.state = HALT
end

function add(computer::Computer, in1_pos::Int, in2_pos::Int, out_pos::Int)
    in1 = get!(computer.ram, in1_pos, 0)
    in2 = get!(computer.ram, in2_pos, 0)
    computer.ram[out_pos] = in1 + in2
    computer.instruction_pointer += 4
end

function multiply(computer::Computer, in1_pos::Int, in2_pos::Int, out_pos::Int)
    in1 = get!(computer.ram, in1_pos, 0)
    in2 = get!(computer.ram, in2_pos, 0)
    computer.ram[out_pos] = in1 * in2
    computer.instruction_pointer += 4
end

function inp(computer::Computer, in_pos::Int)
    computer.ram[in_pos] = popfirst!(computer.inputs)
    computer.instruction_pointer += 2
end

function output(computer::Computer, in_pos::Int)
    push!(computer.outputs, get!(computer.ram, in_pos, 0))
    computer.instruction_pointer += 2
    computer.state = WAIT
end

function jump_if_true(computer::Computer, in1_pos::Int, in2_pos::Int)
    if get!(computer.ram, in1_pos, 0) != 0
        computer.instruction_pointer = get!(computer.ram, in2_pos, 0)
    else
        computer.instruction_pointer += 3
    end
end

function jump_if_false(computer::Computer, in1_pos::Int, in2_pos::Int)
    if get!(computer.ram, in1_pos, 0) == 0
        computer.instruction_pointer = get!(computer.ram, in2_pos, 0)
    else
        computer.instruction_pointer += 3
    end
end

function less_than(computer::Computer, in1_pos::Int, in2_pos::Int, out_pos::Int)
    if get!(computer.ram, in1_pos, 0) < get!(computer.ram, in2_pos, 0)
        computer.ram[out_pos] = 1
    else
        computer.ram[out_pos] = 0
    end
    computer.instruction_pointer += 4
end

function equals(computer::Computer, in1_pos::Int, in2_pos::Int, out_pos::Int)
    if get!(computer.ram, in1_pos, 0) == get!(computer.ram, in2_pos, 0)
        computer.ram[out_pos] = 1
    else
        computer.ram[out_pos] = 0
    end
    computer.instruction_pointer += 4
end

function offset_relative_base(computer::Computer, in_pos::Int)
    computer.relative_base = get!(computer.ram, in_pos, 0) + computer.relative_base
    computer.instruction_pointer += 2
end
end

File: 2019/src/intcode.rs
use std::{collections::HashMap, io};

use anyhow::Result;
use thiserror::Error;

#[derive(Debug, Clone)]
struct Ram {
    memory: HashMap<usize, i64>,
}

impl Ram {
    fn new() -> Self {
        Ram {
            memory: HashMap::new(),
        }
    }

    fn set_memory(&mut self, memory: HashMap<usize, i64>) {
        self.memory = memory;
    }

    fn set_at(&mut self, at: usize, value: i64) {
        self.memory.insert(at, value);
    }

    fn read_at(&self, at: &usize) -> i64 {
        *self.memory.get(at).unwrap_or(&0)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum State {
    Idle,
    Run,
    AdvanceBy(usize),
    MoveTo(usize),
    Halt,
}

#[derive(Debug, Clone, Copy)]
enum Parameter {
    ReadPosition(usize),
    ReadImmediate(i64),
    WritePosition(i64),
}

impl Parameter {
    fn parse(mode: i64, at: usize, read: bool, ram: &Ram) -> Result<Self, InstructionParseError> {
        match read {
            true => match mode {
                0 => Ok(Self::ReadPosition(ram.read_at(&at) as usize)),
                1 => Ok(Self::ReadImmediate(ram.read_at(&at))),
                _ => Err(InstructionParseError::InvalidParameterMode { mode, position: at }),
            },
            false => match mode {
                0 => Ok(Self::WritePosition(ram.read_at(&at))),
                _ => Err(InstructionParseError::InvalidParameterMode { mode, position: at }),
            },
        }
    }

    fn get(&self, ram: &Ram) -> i64 {
        match *self {
            Self::ReadPosition(at) => ram.read_at(&at),
            Self::ReadImmediate(value) => value,
            Self::WritePosition(at) => at,
        }
    }
}

#[derive(Debug, Clone, Copy)]
enum Instruction {
    Halt, // 99
    Add {
        op1: Parameter,
        op2: Parameter,
        out_pos: Parameter,
    }, // 1
    Multiply {
        op1: Parameter,
        op2: Parameter,
        out_pos: Parameter,
    }, // 2
    StoreInput {
        out_pos: Parameter,
    }, // 3
    GiveOutput {
        read_pos: Parameter,
    }, // 4
    JumpIfTrue {
        op1: Parameter,
        jump_pos: Parameter,
    }, // 5
    JumpIfFalse {
        op1: Parameter,
        jump_pos: Parameter,
    }, // 6
    LessThan {
        op1: Parameter,
        op2: Parameter,
        out_pos: Parameter,
    }, // 7
    Equals {
        op1: Parameter,
        op2: Parameter,
        out_pos: Parameter,
    }, // 8
}

#[derive(Debug, Error)]
enum InstructionParseError {
    #[error("invalid opcode {opcode:?} at position {position:?}")]
    InvalidOpcode { opcode: i64, position: usize },
    #[error("invalid parameter mode {mode:?} at position {position:?}")]
    InvalidParameterMode { mode: i64, position: usize },
}

impl Instruction {
    fn parse(ram: &Ram, at: usize) -> Result<Self, InstructionParseError> {
        let instr = ram.read_at(&at);
        let opcode = instr % 100;
        let instr = instr / 100;
        let param1_mode = instr % 10;
        let instr = instr / 10;
        let param2_mode = instr % 10;
        let param3_mode = instr / 10;

        match opcode {
            99 => Ok(Self::Halt),
            1 => Ok(Self::Add {
                op1: Parameter::parse(param1_mode, at + 1, true, ram)?,
                op2: Parameter::parse(param2_mode, at + 2, true, ram)?,
                out_pos: Parameter::parse(param3_mode, at + 3, false, ram)?,
            }),
            2 => Ok(Self::Multiply {
                op1: Parameter::parse(param1_mode, at + 1, true, ram)?,
                op2: Parameter::parse(param2_mode, at + 2, true, ram)?,
                out_pos: Parameter::parse(param3_mode, at + 3, false, ram)?,
            }),
            3 => Ok(Self::StoreInput {
                out_pos: Parameter::parse(param1_mode, at + 1, false, ram)?,
            }),
            4 => Ok(Self::GiveOutput {
                read_pos: Parameter::parse(param1_mode, at + 1, true, ram)?,
            }),
            5 => Ok(Self::JumpIfTrue {
                op1: Parameter::parse(param1_mode, at + 1, true, ram)?,
                jump_pos: Parameter::parse(param2_mode, at + 2, true, ram)?,
            }),
            6 => Ok(Self::JumpIfFalse {
                op1: Parameter::parse(param1_mode, at + 1, true, ram)?,
                jump_pos: Parameter::parse(param2_mode, at + 2, true, ram)?,
            }),
            7 => Ok(Self::LessThan {
                op1: Parameter::parse(param1_mode, at + 1, true, ram)?,
                op2: Parameter::parse(param2_mode, at + 2, true, ram)?,
                out_pos: Parameter::parse(param3_mode, at + 3, false, ram)?,
            }),
            8 => Ok(Self::Equals {
                op1: Parameter::parse(param1_mode, at + 1, true, ram)?,
                op2: Parameter::parse(param2_mode, at + 2, true, ram)?,
                out_pos: Parameter::parse(param3_mode, at + 3, false, ram)?,
            }),
            _ => Err(InstructionParseError::InvalidOpcode {
                opcode,
                position: at,
            }),
        }
    }

    fn execute(&self, ram: &mut Ram) -> State {
        match *self {
            Self::Halt => State::Halt,
            Self::Add { op1, op2, out_pos } => {
                let out = op1.get(ram) + op2.get(ram);
                ram.set_at(out_pos.get(ram) as usize, out);
                State::AdvanceBy(4)
            }
            Self::Multiply { op1, op2, out_pos } => {
                let out = op1.get(ram) * op2.get(ram);
                ram.set_at(out_pos.get(ram) as usize, out);
                State::AdvanceBy(4)
            }
            Self::StoreInput { out_pos } => {
                println!("Give input");
                let mut buf = String::new();
                io::stdin()
                    .read_line(&mut buf)
                    .expect("error reading input");
                let value: i64 = buf.trim().parse().expect("input is not an integer");
                ram.set_at(out_pos.get(ram) as usize, value);
                State::AdvanceBy(2)
            }
            Self::GiveOutput { read_pos } => {
                println!("{}", read_pos.get(ram));
                State::AdvanceBy(2)
            }
            Self::JumpIfTrue { op1, jump_pos } => {
                if op1.get(ram) != 0 {
                    State::MoveTo(jump_pos.get(ram) as usize)
                } else {
                    State::AdvanceBy(3)
                }
            }
            Self::JumpIfFalse { op1, jump_pos } => {
                if op1.get(ram) == 0 {
                    State::MoveTo(jump_pos.get(ram) as usize)
                } else {
                    State::AdvanceBy(3)
                }
            }
            Self::LessThan { op1, op2, out_pos } => {
                let at = out_pos.get(ram) as usize;
                if op1.get(ram) < op2.get(ram) {
                    ram.set_at(at, 1);
                } else {
                    ram.set_at(at, 0);
                }
                State::AdvanceBy(4)
            }
            Self::Equals { op1, op2, out_pos } => {
                let at = out_pos.get(ram) as usize;
                if op1.get(ram) == op2.get(ram) {
                    ram.set_at(at, 1);
                } else {
                    ram.set_at(at, 0);
                }
                State::AdvanceBy(4)
            }
        }
    }
}

pub struct Computer {
    ram: Ram,
    state: State,
    instruction_pointer: usize,
}

impl Computer {
    pub fn new() -> Self {
        Computer {
            ram: Ram::new(),
            state: State::Idle,
            instruction_pointer: 0,
        }
    }

    pub fn initialize(&mut self) {
        self.state = State::Run;
    }

    pub fn load_program(&mut self, program: HashMap<usize, i64>) {
        self.ram.set_memory(program);
        self.instruction_pointer = 0;
    }

    pub fn read_memory(&self, at: &usize) -> i64 {
        self.ram.read_at(at)
    }

    pub fn run(&mut self) -> Result<()> {
        while self.state != State::Halt {
            let instruction = Instruction::parse(&self.ram, self.instruction_pointer)?;
            // println!("{instruction:?}");
            let new_state = instruction.execute(&mut self.ram);
            match new_state {
                State::Halt => {
                    self.instruction_pointer += 1;
                    self.state = State::Halt;
                }
                State::AdvanceBy(by) => {
                    self.instruction_pointer += by;
                    self.state = State::Run;
                }
                State::MoveTo(position) => {
                    self.instruction_pointer = position;
                    self.state = State::Run;
                }
                State::Idle => self.state = State::Idle,
                State::Run => self.state = State::Run,
            }
        }
        Ok(())
    }
}

impl Default for Computer {
    fn default() -> Self {
        Self::new()
    }
}
